text/0000-type-declarations.md

```diff
@@ -0,0 +1,219 @@
+- Start Date: 2018-02-21
+- RFC PR:
+- WordPress Coding Standards Issue:
+
+# Summary
+
+[Type Declarations](http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration) make code more readable. They also protect code, because Type Declarations are a way of enforcing specific data types in parameters passed to a function.
+
+# Basic Example
+
+You can force callers to pass an `array`, or a specific type of object, such as a `WP_REST_Request`. If a caller doesn't pass the data type expected, PHP 5 triggers a recoverable fatal error and PHP 7 throws a [TypeError](http://php.net/manual/en/class.typeerror.php) exception. This behavior catches problems right away.
+
+```php
+protected function foo( Bar $bar, array $items ) {
+	// $bar   must be an instance of the Bar class.
+	// $items must be passed as an array.
+}
+```
+
+# Motivation
+
+PHP is a [weakly typed language](https://en.wikipedia.org/wiki/Strong_and_weak_typing). It doesn't require you to declare data types. However, variables still have data types associated with them (e.g., `string`, `int`). In a weakly typed language you can do things like adding a `string` to an `int` with no error. While this can be wonderful in some cases, it can also lead to unanticipated behavior and bugs.
+
+For example, PHP functions and class methods accept parameters, such as `function foo( $bar, $items )`. If Type Declarations aren't being used, the only way a function knows what it's being passed is to run `is_*()` checks, `instanceof`, or use type casting. Otherwise, `$bar` could be anything, and there's a chance that invalid data types would go unnoticed and produce an unexpected or incorrect return value. Imagine type casting an array to an integer. That's forcing a wrong into a right, instead of correcting the underlying issue. A caller should pass the right data type to begin with.
+
```

Comment:
I'd add an additional paragraph here along the lines of:
```
Moreover, as the wrong value does not throw an immediate error, it might get passed down to another function,
and yet another, and so on. Most often, you'll end up with an error that is thrown in a completely unrelated
part of the code, or in the worst case, no error being thrown at all, while the frontend happily renders bogus
data.
```
Basically emphasizing why throwing an immediate error is so important.